File: obsidian-plugin/src/config/diaryConfig.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";

export const DIARY_PERIODS = [
  "daily",
  "weekly",
  "monthly",
  "quarterly",
  "yearly",
] as const;

export type DiaryPeriod = (typeof DIARY_PERIODS)[number];

export interface DiarySettings {
  rootPath: string;
  templatePaths: Record<DiaryPeriod, string>;
}

export const DEFAULT_DIARY_SETTINGS: DiarySettings = {
  rootPath: "Journal",
  templatePaths: {
    daily: ".crabby/templates/diary/daily.md",
    weekly: ".crabby/templates/diary/weekly.md",
    monthly: ".crabby/templates/diary/monthly.md",
    quarterly: ".crabby/templates/diary/quarterly.md",
    yearly: ".crabby/templates/diary/yearly.md",
  },
};

export function normalizeDiarySettings(source: unknown): DiarySettings {
  const record = isRecord(source) ? source : {};
  const rootPath = normalizeVaultRelativePath(
    record.rootPath,
    DEFAULT_DIARY_SETTINGS.rootPath,
    "rootPath",
  );
  const rawTemplatePaths = isRecord(record.templatePaths)
    ? record.templatePaths
    : {};

  return {
    rootPath,
    templatePaths: {
      daily: normalizeVaultRelativePath(
        rawTemplatePaths.daily,
        DEFAULT_DIARY_SETTINGS.templatePaths.daily,
        "templatePaths.daily",
      ),
      weekly: normalizeVaultRelativePath(
        rawTemplatePaths.weekly,
        DEFAULT_DIARY_SETTINGS.templatePaths.weekly,
        "templatePaths.weekly",
      ),
      monthly: normalizeVaultRelativePath(
        rawTemplatePaths.monthly,
        DEFAULT_DIARY_SETTINGS.templatePaths.monthly,
        "templatePaths.monthly",
      ),
      quarterly: normalizeVaultRelativePath(
        rawTemplatePaths.quarterly,
        DEFAULT_DIARY_SETTINGS.templatePaths.quarterly,
        "templatePaths.quarterly",
      ),
      yearly: normalizeVaultRelativePath(
        rawTemplatePaths.yearly,
        DEFAULT_DIARY_SETTINGS.templatePaths.yearly,
        "templatePaths.yearly",
      ),
    },
  };
}

export function serializeDiarySettings(settings: DiarySettings): Record<string, unknown> {
  return {
    rootPath: settings.rootPath,
    templatePaths: { ...settings.templatePaths },
  };
}

export function readDiarySettingsFile(filePath: string): DiarySettings | null {
  if (!existsSync(filePath)) {
    return null;
  }
  try {
    const raw = JSON.parse(readFileSync(filePath, "utf8")) as unknown;
    return normalizeDiarySettings(raw);
  } catch {
    return null;
  }
}

export function writeDiarySettingsFile(filePath: string, settings: DiarySettings): void {
  mkdirSync(dirname(filePath), { recursive: true });
  writeFileSync(
    filePath,
    `${JSON.stringify(serializeDiarySettings(settings), null, 2)}\n`,
    "utf8",
  );
}

export function normalizeVaultRelativePath(
  value: unknown,
  fallback: string,
  fieldName: string,
): string {
  const raw = typeof value === "string" ? value.trim() : "";
  const source = raw || fallback;
  const normalized = source.replace(/\\/g, "/").trim();

  if (
    normalized.startsWith("/") ||
    normalized.startsWith("~") ||
    /^[A-Za-z]:/.test(normalized)
  ) {
    throw new Error(`${fieldName} 必须是 Vault-relative 路径。`);
  }

  const segments = normalized
    .split("/")
    .filter((segment) => segment && segment !== ".");
  if (segments.some((segment) => segment === "..")) {
    throw new Error(`${fieldName} 不能包含 ".."。`);
  }

  return segments.join("/") || fallback;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

export function resolveDiaryConfigPath(vaultRoot: string): string {
  return resolve(vaultRoot, ".crabby", "config", "diary.json");
}

File: server/diary_config.py
"""Diary configuration shared by backend tools and plugin-written config."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from typing import Any

DIARY_PERIODS = ("daily", "weekly", "monthly", "quarterly", "yearly")
DEFAULT_DIARY_ROOT_PATH = "Journal"
DEFAULT_DIARY_TEMPLATE_PATHS: dict[str, str] = {
    "daily": ".crabby/templates/diary/daily.md",
    "weekly": ".crabby/templates/diary/weekly.md",
    "monthly": ".crabby/templates/diary/monthly.md",
    "quarterly": ".crabby/templates/diary/quarterly.md",
    "yearly": ".crabby/templates/diary/yearly.md",
}
DIARY_CONFIG_RELATIVE_PATH = Path(".crabby") / "config" / "diary.json"

_WINDOWS_DRIVE_RE = re.compile(r"^[A-Za-z]:")


class DiaryConfigError(ValueError):
    """Raised when diary configuration is malformed or unsafe."""


@dataclass(frozen=True)
class DiaryConfig:
    root_path: str
    template_paths: dict[str, str]

    def to_json_dict(self) -> dict[str, Any]:
        return {
            "rootPath": self.root_path,
            "templatePaths": {
                period: self.template_paths[period]
                for period in DIARY_PERIODS
            },
        }


def default_diary_config() -> DiaryConfig:
    return DiaryConfig(
        root_path=DEFAULT_DIARY_ROOT_PATH,
        template_paths=dict(DEFAULT_DIARY_TEMPLATE_PATHS),
    )


def diary_config_path(vault_path: Path) -> Path:
    return Path(vault_path).expanduser().resolve() / DIARY_CONFIG_RELATIVE_PATH


def load_diary_config(vault_path: Path) -> DiaryConfig:
    path = diary_config_path(vault_path)
    if not path.is_file():
        return default_diary_config()

    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise DiaryConfigError(f"日记配置 JSON 无法解析: {path}") from exc
    except OSError as exc:
        raise DiaryConfigError(f"日记配置无法读取: {path}") from exc

    if not isinstance(raw, dict):
        raise DiaryConfigError("日记配置必须是 JSON 对象。")

    defaults = default_diary_config()
    root_path = normalize_vault_relative_path(
        raw.get("rootPath"),
        fallback=defaults.root_path,
        field_name="rootPath",
    )

    raw_templates = raw.get("templatePaths")
    if raw_templates is not None and not isinstance(raw_templates, dict):
        raise DiaryConfigError("templatePaths 必须是 JSON 对象。")

    template_paths: dict[str, str] = {}
    for period in DIARY_PERIODS:
        raw_value = raw_templates.get(period) if isinstance(raw_templates, dict) else None
        template_paths[period] = normalize_vault_relative_path(
            raw_value,
            fallback=defaults.template_paths[period],
            field_name=f"templatePaths.{period}",
        )

    return DiaryConfig(root_path=root_path, template_paths=template_paths)


def normalize_vault_relative_path(
    value: object,
    *,
    fallback: str,
    field_name: str,
) -> str:
    raw = str(value).strip() if isinstance(value, str) else ""
    if not raw:
        raw = fallback
    raw = raw.replace("\\", "/").strip()

    if raw.startswith("/") or raw.startswith("~") or _WINDOWS_DRIVE_RE.match(raw):
        raise DiaryConfigError(f"{field_name} 必须是 Vault-relative 路径。")

    path = PurePosixPath(raw)
    if any(part == ".." for part in path.parts):
        raise DiaryConfigError(f"{field_name} 不能包含 '..'。")

    normalized = path.as_posix().strip("/")
    if not normalized or normalized == ".":
        normalized = fallback
    return normalized


def resolve_vault_relative_path(vault_path: Path, relative_path: str) -> Path:
    vault = Path(vault_path).expanduser().resolve()
    target = (vault / relative_path).resolve()
    try:
        target.relative_to(vault)
    except ValueError as exc:
        raise DiaryConfigError(f"路径逃逸出 Vault: {relative_path}") from exc
    return target


def write_default_diary_config_if_missing(vault_path: Path) -> Path:
    path = diary_config_path(vault_path)
    if path.exists():
        return path
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8", newline="\n") as handle:
        handle.write(
            json.dumps(
                default_diary_config().to_json_dict(),
                ensure_ascii=False,
                indent=2,
            )
            + "\n",
        )
    return path
